### PR TITLE
Don't close entire HTTP/2 connection when request is cancelled

### PR DESCRIPTION
**Pull request by idelpivnitskiy**
Motivation:

`LoadBalancedStreamingHttpClient` pessimistically assumes the connection
that serves a request should be closed on request cancellation. However,
in case of HTTP/2 it closes the parent TCP connection instead of a stream
(because stream is not accessible from the connection API). It also fails
all other in-flight requests on the same connection.

Modifications:

- In case of HTTP/2 and above mark request as finished instead of closing
the connection when cancellation happens;
- Test that request cancellation doesn't close parent TCP connection for
HTTP/2 protocol;
- Change `HttpConnectionContext#protocol()` to return `HttpProtocolVersion`;
- Change `GrpcConnectionContext.GrpcProtocol` to return `HttpProtocolVersion`;

Result:

HTTP/2 request cancellation doesn't close the TCP connection.

### CODE REVIEW

File: servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/H2ResponseCancelTest.java [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+        firstRequestReceivedLatch.await();
+
+        AtomicReference<Cancellable> cancellable = new AtomicReference<>();
+        stepVerifier(requester.request(defaultStrategy(), newRequest(requester, "second"))
+                        .whenOnSuccess(responses::add)) // Add response to the queue to verify that we never receive it
+                .expectCancellableConsumed(cancellable::set)
+                .then(() -> {
+                    try {
+                        secondRequestReceivedLatch.await();
+                        cancellable.get().cancel(); // If I use thenCancel() the current then(Runnable) does not run
```
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/H2ResponseCancelTest.java:166]: @Scottmitch if I do this way:
```java
stepVerifier(requester.request(defaultStrategy(), newRequest(requester, "second"))
                .whenOnSuccess(responses::add)) // Add response to the queue to verify that we never receive it
        .expectCancellable()
        .then(() -> {
            try {
                secondRequestReceivedLatch.await();
            } catch (InterruptedException e) {
                // ignore
            }
        })
        .thenCancel()
        .verify();
```
`then(Runnable)` is not invoked, after `expectCancellable()` the next step will be `thenCancel()`. Is this expected behavior?

Comment (Scottmitch) [reply servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/H2ResponseCancelTest.java:166]: discussed offline, we will investigate clarifying/improving as part of followup issues. tldr; the "then" runnable operations are run on the verifier thread, the expectations are run on the subscriber thread (including `thenCancel`), so these aren't sequential.

File: servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/H2ResponseCancelTest.java [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+                    } catch (InterruptedException e) {
+                        // ignore
+                    }
+                })
+                .expectError()
+                // .thenCancel()
+                .verify();
+
+        assertThat("Unexpected responses", responses, hasSize(0));
```
Comment (Scottmitch) [inline servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/H2ResponseCancelTest.java:175]: nit: `hasSize(0)` -> `is(empty())`

File: servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/H2ResponseCancelTest.java [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        firstResponseLatch.countDown();
+
+        assertResponse(responses.take(), HTTP_2_0, OK, "first");
+        // Mak sure we can use the same connection for future requests:
```
Comment (Scottmitch) [inline servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/H2ResponseCancelTest.java:179]: nit: `Mak` -> `Make`

File: servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/H2ResponseCancelTest.java [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+    public void testClient() throws Exception {
+        // Release existing connection to avoid creating unexpected number of new h2 connections because of randomness
+        // in RRLB (it may randomly pick the reserved connection until all attempts are taken).
+        ((ReservedStreamingHttpConnection) streamingHttpConnection()).releaseAsync().toFuture().get();
+
+        int connectionsBefore = newConnectionsCounter.get();
+        requestCancellationResetsStreamButNotParentConnection(streamingHttpClient());
+        int connectionsAfter = newConnectionsCounter.get();
+        assertThat("Client unexpectedly created more connections instead of reusing the existing one",
+                connectionsAfter - connectionsBefore, is(0));
```
Comment (Scottmitch) [inline servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/H2ResponseCancelTest.java:124]: nit: `assertThat(connectionsAfter, is(connectionsBefore));` or `assertEquals(connectionsBefore, connectionsAfter)` will give a more informative error message if a failure occurs.

**Review by Scottmitch — APPROVED** [review]

**Pull request merged**